src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -2095,6 +2167,326 @@ HeapWord* ShenandoahFreeSet::allocate(ShenandoahAllocRequest& req, bool& in_new_
   }
 }
 
+HeapWord* ShenandoahFreeSet::allocate_humongous(ShenandoahAllocRequest& req) {
+  assert(ShenandoahHeapRegion::requires_humongous(req.size()), "Must be humongous alloc");
+  ShenandoahHeapLocker locker(_heap->lock(), req.is_mutator_alloc());
+  return allocate_contiguous(req, /*is_humongous*/true);
+}
+
+HeapWord* ShenandoahFreeSet::allocate_contiguous_cds(ShenandoahAllocRequest& req) {
+  assert(req.type() == ShenandoahAllocRequest::_alloc_cds, "Must be CDS alloc.");
+  ShenandoahHeapLocker locker(_heap->lock(), req.is_mutator_alloc());
+  return allocate_contiguous(req, /*is_humongous*/false);
+}
+
+template<bool IS_TLAB>
+HeapWord* ShenandoahFreeSet::cas_allocate_single_for_mutator(
+  uint probe_start, uint probe_count, ShenandoahAllocRequest &req, bool &in_new_region) {
+  HeapWord *obj = nullptr;
+  uint i = 0u;
+  while (i < probe_count) {
+    {
+      // Yield to Safepoint
+      ThreadBlockInVM tbivm(JavaThread::current());
+    }
+    uint idx = (probe_start + i) % ShenandoahDirectlyAllocatableRegionCount;
+    ShenandoahDirectAllocationRegion& shared_region = _direct_allocation_regions[idx];
+    ShenandoahHeapRegion* r = nullptr;
+    // Intentionally not using AtomicAccess::load, if a mutator see a stale region it will fail to allocate anyway.
+    if ((r = shared_region._address) != nullptr && r->reserved_for_direct_allocation()) {
+      obj = cas_allocate_in_for_mutator<IS_TLAB>(r, req, in_new_region);
+      if (obj != nullptr) {
+        return obj;
+      }
+    }
+    i++;
+  }
+  return obj;
+}
+template<bool IS_TLAB>
+HeapWord* ShenandoahFreeSet::try_allocate_single_for_mutator(ShenandoahAllocRequest &req, bool &in_new_region) {
+  shenandoah_assert_not_heaplocked();
+  assert(req.is_mutator_alloc(), "Must be mutator allocation");
+  assert(req.is_young(), "Mutator allocations always come from young generation.");
+  assert(!ShenandoahHeapRegion::requires_humongous(req.size()), "Must not");
+  assert(req.type() == ShenandoahAllocRequest::_alloc_tlab ||
+         req.type() == ShenandoahAllocRequest::_alloc_shared ||
+         req.type() == ShenandoahAllocRequest::_alloc_cds,  "Must be");
+
+  uint start_idx = ShenandoahDirectlyAllocatableRegionAffinity::index();
+  const uint max_probes = ShenandoahDirectAllocationMaxProbes;
+  for (;;) {
+    HeapWord* obj = nullptr;
```

Comment:
This is written as an infinite loop, but I'm not sure it intends to iterate?  It looks like there's no way for the loop body to get to a second iteration.